python/feathub/feature_views/feature_view.py

```diff
@@ -192,3 +205,38 @@ def get_bounded_view(self) -> TableDescriptor:
         feature_view = deepcopy(self)
         feature_view.source = self.get_resolved_source().get_bounded_view()
         return feature_view
+
+    def _derive_feature_dtype(self, feature: Feature) -> Optional[DType]:
+        transform = feature.transform
+        if isinstance(transform, ExpressionTransform):
+            return self._derive_feature_dtype_from_feathub_expression(transform.expr)
+        elif isinstance(transform, OverWindowTransform):
+            expr_result_type = self._derive_feature_dtype_from_feathub_expression(
+                transform.expr
+            )
+            return transform.agg_func.get_result_type(expr_result_type)
+        elif isinstance(transform, SlidingWindowTransform):
+            expr_result_type = self._derive_feature_dtype_from_feathub_expression(
+                transform.expr
+            )
+            return transform.agg_func.get_result_type(expr_result_type)
+        elif isinstance(transform, JoinTransform):
+            # We don't need to derive data type of the feature to join. Its data type is
```

Comment:
We need to get the dtype of the feature to join to determine the dype of the output feature, right?

We should be able to know the dtype of every feature. If the program references a feature on another feature view in the feature registry, we should be able to retrieve the dtype of this feature from the feature registry. This is done in `_DerivedFeatureView#get_feature_from_feature_str`. In general, the dtype of the feature can be obtained when we resolve the feature view.

Reply:
You are right that we should know the data type of the join feature. Actually, this method should not be called with the JoinTransform feature, as its data should not be None. So I think it is better to throw an exception here.